### PR TITLE
Upload-CreditReport FINERACT-1285

### PR DESCRIPTION
**Pull request by rrpawar96**
## Description

Uploading Credit Reports back to ThitsaWorks Credit Bureau, https://issues.apache.org/jira/browse/FINERACT-1285.

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by vorburger** [comment]
@rrpawar96 FYI this has SpotBugs violations you'll want to fix.

Also, you may be affected by FINERACT-924 after you fix SpotBugs, so rebase after that is fixed.

**Comment by rrpawar96** [comment]
@vorburger fixed the build and Ready for review. 
Also have updated the documentation (uploading the creditReport): https://cwiki.apache.org/confluence/display/FINERACT/Documentation+to+use+Integrated+Credit+Bureau

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/creditbureau/service/ThitsaWorksCreditBureauIntegrationWritePlatformServiceImpl.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/creditbureau/service/ThitsaWorksCreditBureauIntegrationWritePlatformServiceImpl.java]: @rrpawar96 did you mean to change this, or this a left over from a local test that you did not mean to change? If it should change, perhaps you would prefer to just remove instead of comment out the bean validation which you don't seem to want to do here anymore? (But why?)

**Review by vorburger — APPROVED** [review]
LGTM, have 1 minor feedback. @rrpawar96 should @nikpawar89 review and LGTM this as well? Or shall we merge it, after you commented on that minor feedback? Tx.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/creditbureau/service/ThitsaWorksCreditBureauIntegrationWritePlatformServiceImpl.java
Comment (rrpawar96) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/creditbureau/service/ThitsaWorksCreditBureauIntegrationWritePlatformServiceImpl.java]: Yes, it was the change in the implementation of the type of Exception, (PlatformDataIntegrityException instead of PlatformApiDataValidationException.) PlatformDataIntegrityException was more suitable to throw this exception.

**Comment by rrpawar96** [comment]
> LGTM, have 1 minor feedback. @rrpawar96 should @nikpawar89 review and LGTM this as well? Or shall we merge it, after you commented on that minor feedback? Tx.

Yes, would like to get LGTM from @nikpawar89 too.

**Comment by nikpawar89** [comment]
@rrpawar96 This looks fine to me but I would expect report generation to be automated. It would be difficult for users to generate report manually, I think we should automate it.

We have different options here : 
1) Create a job, which generates the file and uploads to Credit Bureau.
2) User triggers the report generation - looks into the report then submits to Credit Bureau.

You may have both features  or start with second option, build trust and then automate the process.

Let's take step by step approach, let us implement second option and then option one should be easy.
Regarding report generation, You would have to identify the fields in the schema to be mapped to fields of the template provided by Credit Bureau.  Also consider, the row limitation of file format (or constraints from credit bureau), if there is any.

You may also want to take a more generic approach, where in user has the ability to create/update mapping, format etc - sky is the limit :)

**For now, the upload feature looks good to me. Report generation can be worked independently.**

**Comment by rrpawar96** [comment]
@nikpawar89, thanks for the review, and for giving the idea of generating the report by automation and also thanks for the explanation on how to approach the same, will create a JIRA for this enhancement.

**Comment by vorburger** [comment]
@nikpawar89 sorry just to make this more explicit, does this have LGTM Approval from you for merge? Will you merge it, or do you want me to? Just asking to avoid confusion. Thanks!

**Comment by nikpawar89** [comment]
> @nikpawar89 sorry just to make this more explicit, does this have LGTM Approval from you for merge? Will you merge it, or do you want me to? Just asking to avoid confusion. Thanks!

@vorburger I gave LGTM, unfortunately I don't have access to merge.

**Comment by vorburger** [comment]
> @vorburger I gave LGTM, unfortunately I don't have access to merge.

oh OK sorry I forgot; I'll merge it. If you [contribute more yourself](https://github.com/apache/fineract/commits?author=nikpawar89), you can earn the committer privilege!

**Comment by nikpawar89** [comment]
> > @vorburger I gave LGTM, unfortunately I don't have access to merge.
> 
> 
> 
> oh OK sorry I forgot; I'll merge it. If you [contribute more yourself](https://github.com/apache/fineract/commits?author=nikpawar89), you can earn the committer privilege!

I am already a committer- but don't know how would I get merge access.

**Pull request merged**
